grovedb/src/batch/mod.rs

```diff
@@ -2517,177 +2407,93 @@
-            batch_apply_options.batch_pause_height = None;
+        let continue_storage_batch = StorageBatch::new();
 
-            let continue_storage_batch = StorageBatch::new();
+        cost_return_on_error!(
+            &mut cost,
+            self.continue_partial_apply_body(
+                left_over_operations,
+                new_operations,
+                Some(batch_apply_options),
+                update_element_flags_function,
+                split_removal_bytes_function,
+                |path, new_merk| {
+                    self.open_batch_transactional_merk_at_path(
+                        &continue_storage_batch,
+                        path.into(),
+                        tx.as_ref(),
+                        new_merk,
+                        grove_version,
+                    )
+                },
+                grove_version
+            )
+        );
 
-            cost_return_on_error!(
-                &mut cost,
-                self.continue_partial_apply_body(
-                    left_over_operations,
-                    new_operations,
-                    Some(batch_apply_options),
-                    update_element_flags_function,
-                    split_removal_bytes_function,
-                    |path, new_merk| {
-                        self.open_batch_merk_at_path(
-                            &continue_storage_batch,
-                            path.into(),
-                            new_merk,
-                            grove_version,
-                        )
-                    },
-                    grove_version
-                )
-            );
+        // let's build the write batch
+        let continued_pending_costs = cost_return_on_error!(
+            &mut cost,
+            self.db
+                .continue_write_batch(&mut write_batch, continue_storage_batch)
+                .map_err(|e| e.into())
+        );
 
-            // let's build the write batch
-            let continued_pending_costs = cost_return_on_error!(
-                &mut cost,
-                self.db
-                    .continue_write_batch(&mut write_batch, continue_storage_batch)
-                    .map_err(|e| e.into())
-            );
+        pending_costs.add_assign(continued_pending_costs);
 
-            pending_costs.add_assign(continued_pending_costs);
+        // TODO: compute batch costs
+        cost_return_on_error!(
+            &mut cost,
+            self.db
+                .commit_db_write_batch(write_batch, pending_costs, Some(tx.as_ref()))
+                .map_err(|e| e.into())
+        );
 
-            // TODO: compute batch costs
-            cost_return_on_error!(
-                &mut cost,
-                self.db
-                    .commit_db_write_batch(write_batch, pending_costs, None)
-                    .map_err(|e| e.into())
-            );
-        }
-        Ok(()).wrap_with_cost(cost)
+        tx.commit_local().wrap_with_cost(cost)
```

Comment:
What are you talking about, we have one batch commit in the end as well as committing transaction only in case we opened it when it wasn't passed from above